packages/e2e-tests/helpers/source-panel.ts

```diff
@@ -12,26 +12,72 @@ export async function addBreakpoint(
     columnIndex?: number;
     lineNumber: number;
     url: string;
+    isNode?: boolean;
   }
 ): Promise<void> {
-  const { lineNumber, url } = options;
+  const { lineNumber, url, isNode = false } = options;
 
   debugPrint(`Adding breakpoint at ${chalk.bold(`${url}:${lineNumber}`)}`, "addBreakpoint");
 
-  await openDevToolsTab(page);
+  if (!isNode) {
+    await openDevToolsTab(page);
+  }
 
   if (url) {
     await openSourceExplorerPanel(page);
     await openSource(page, url);
   }
 
+  await scrollUntilLineIsVisible(page, lineNumber);
+
   const line = await getSourceLine(page, lineNumber);
   await line.locator(".CodeMirror-linenumber").hover();
   await line.locator(".CodeMirror-linenumber").click();
 
   await waitForBreakpoint(page, options);
 }
 
+async function scrollUntilLineIsVisible(page: Page, lineNumber: number) {
+  let loopCounter = 0;
+
+  while (true) {
+    const visibleLineNumbers = await getVisibleLineNumbers(page);
+
+    if (visibleLineNumbers.includes(lineNumber)) {
+      break;
+    } else {
+      if (loopCounter > 10) {
+        throw new Error("Stuck in infinite scrolling loop looking for  line: " + lineNumber);
+      }
+    }
+    // it must not be on screen, we have to scroll
+    const [first] = visibleLineNumbers;
+
+    const scrollUp = lineNumber < first;
+    const deltaY = scrollUp ? -1000 : 1000;
+    const scroller = page.locator(".CodeMirror-scroll").first();
+    await scroller.click();
+    await page.mouse.wheel(0, deltaY);
+    loopCounter++;
+  }
+}
+
+async function getVisibleLineNumbers(page: Page) {
+  const lineNumberElements = page.locator(".CodeMirror-linenumber");
+  const numElements = await lineNumberElements.count();
+
+  const textContents = await lineNumberElements.allTextContents();
+  const allLineNumbers = textContents.map(s => Number(s));
+
+  const isLine1Visible = allLineNumbers.includes(1);
+  const firstSliceIndex = isLine1Visible ? 0 : 9;
+  // CodeMirror normally has a 10-line buffer on either side of the viewport.
+  // Limit to just what's actually on screen
+  const lineNumbers = allLineNumbers.slice(firstSliceIndex, allLineNumbers.length - 10);
```

Comment:
`isVisible` won't be sufficient, for the same reasons that the side panel checks don't work.  Playwright returns `true` for `isVisible()` if the content exists in the DOM and has a non-zero size.  But, in this case, the problem is that the elements _do_ exist in the DOM... but _outside_ the visible range.

I think CodeMirror has an option to configure its overscroll buffer size, but I don't think this would change unless we specifically modify that value.